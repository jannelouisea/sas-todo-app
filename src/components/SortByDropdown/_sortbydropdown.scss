.sort-by-dropdown {
    display: flex;
    flex-direction: column;
    align-items: flex-start;

    @media (min-width: 768px) {
        align-items: flex-end;
    }
}