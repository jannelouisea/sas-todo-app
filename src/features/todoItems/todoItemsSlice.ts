import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import {
    IToDoItem
} from 'src/interfaces'
import moment from 'moment';
import {
    Sort
} from 'src/static/enums'

export type ITodoItemAddedPayload = {
    id: string,
    text: string,
    createdAt: number,
};

const initialState: IToDoItem[] = [
    {
        id: 'init-item-001',
        text: 'Clean the house',
        createdAt: moment().unix(),
        completed: false
    },
    {
        id: 'init-item-002',
        text: 'Pack things for upcoming trip',
        createdAt: 1689652800, // Tue Jul 18 2023 00:00:00 GMT-0400 (EST)
        completed: false
    },
    {
        id: 'init-item-003',
        text: 'Make dinner',
        createdAt: 1690257600, // Tue Jul 25 2023 00:00:00 GMT-0400 (EST)
        completed: true
    },
    {
        id: 'init-item-004',
        text: 'Prepare for SAS Interview',
        createdAt: 1689393600, // Sat Jul 15 2023 00:00:00 GMT-0400 (EST)

        completed: true
    },
];

const todoItemsSlice = createSlice({
    name: 'todoItems',
    initialState,
    reducers: {
        todoItemAdded(state, action: PayloadAction<ITodoItemAddedPayload>) {
            const { id, text, createdAt } = action.payload;
            const newItem = {
                id,
                text,
                createdAt,
                completed: false
            };

            return [newItem, ...state];
        },
        todoItemToggled(state, action: PayloadAction<string>) {
            return state.map(item => {
                if (item.id !== action.payload) {
                    return item
                }

                return {
                    ...item,
                    completed: !item.completed
                }
            });
        },
        todoItemRemoved(state, action: PayloadAction<string>) {
            return state.filter(item => item.id !== action.payload);
        },
        todoItemsSorted(state, action: PayloadAction<Sort>) {
            let newState = [...state];

            switch (action.payload) {
                case Sort.DateDesc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemB.createdAt - itemA.createdAt
                        //moment(itemB.createdAt).format('YYYYMMDD').localeCompare(moment(itemA.createdAt).format('YYYYMMDD'))
                    );
                case Sort.DateAsc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemA.createdAt - itemB.createdAt
                        //moment(itemA.createdAt).format('YYYYMMDD').localeCompare(moment(itemB.createdAt).format('YYYYMMDD'))
                    );
                case Sort.TextDesc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemB.text.localeCompare(itemA.text)
                    );
                case Sort.TextAsc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemA.text.localeCompare(itemB.text)
                    );
                case Sort.Cmpl1st:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) => {
                        if (itemA.completed === itemB.completed) {
                            //return moment(itemB.createdAt).format('YYYYMMDD').localeCompare(moment(itemA.createdAt).format('YYYYMMDD'))
                            return itemB.createdAt - itemA.createdAt;
                        } else if (itemA.completed) {
                            return -1;
                        } else {
                            return 1;
                        }
                    });
                case Sort.Incmpl1st:
                default:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) => {
                        if (itemA.completed === itemB.completed) {
                            //return moment(itemB.createdAt).format('YYYYMMDD').localeCompare(moment(itemA.createdAt).format('YYYYMMDD'))
                            return itemB.createdAt - itemA.createdAt;
                        } else if (itemA.completed) {
                            return 1;
                        } else {
                            return -1;
                        }
                    });
            }
        }
    }
});

export const { todoItemAdded, todoItemToggled, todoItemRemoved, todoItemsSorted } = todoItemsSlice.actions;

export default todoItemsSlice.reducer;