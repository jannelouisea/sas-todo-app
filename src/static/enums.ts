export enum Sort {
    Incmpl1st = 'INCMPL_1ST',
    Cmpl1st = 'CMPL_1ST',
    DateAsc = 'DATE_ASC',
    DateDesc = 'DATE_DESC',
    TextAsc = 'TEXT_ASC',
    TextDesc = 'TEXT_DESC',
}

export enum MUIColor {
    Primary = 'primary',
    Success = 'success',
    Warning = 'warning',
    Neutral = 'neutral',
    Danger = 'danger',
}

export enum MUIIconColor {
    Inherit = 'inherit',
}

export enum MUIIconSize {
    Inherit = 'inherit',
}

export enum MUIVariant {
    Solid = 'solid',
    Soft = 'soft',
    Outlined = 'outlined',
    Plain = 'plain',
}

export enum MUISize {
    Small = 'sm',
    Medium = 'md',
    Large = 'lg',
}

export enum MUIPos {
    Right = 'right',
}

export enum MUIText {
    Label = 'body3'
}

export enum MUIFontSize {
    Small = 'small'
}