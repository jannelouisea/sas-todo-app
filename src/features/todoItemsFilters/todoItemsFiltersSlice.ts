import { createSlice, PayloadAction } from '@reduxjs/toolkit'

import { Sort } from 'src/static/enums'
import { ITodoItemsFilters } from 'src/interfaces';

const initialState: ITodoItemsFilters = {
    searchTerm: '',
    showNewer: true,
    showOlder: true,
    showCompleted: true,
    sortBy: Sort.Incmpl1st,
};

const todoItemsFiltersSlice = createSlice({
    name: 'todoItemsFilters',
    initialState,
    reducers: {
        searchTermChanged(state, action: PayloadAction<string>) {
            return { ...state, searchTerm: action.payload }
        },
        showNewerToggled(state) {
            return { ...state, showNewer: !state.showNewer }
        },
        showOlderToggled(state) {
            return { ...state, showOlder: !state.showOlder }
        },
        showCompletedToggled(state) {
            return { ...state, showCompleted: !state.showCompleted }
        },
        sortByChanged(state, action: PayloadAction<Sort>) {
            return { ...state, sortBy: action.payload }
        }
    }
});

export const {
    searchTermChanged,
    showNewerToggled,
    showOlderToggled,
    showCompletedToggled,
    sortByChanged,
} = todoItemsFiltersSlice.actions;

export default todoItemsFiltersSlice.reducer;