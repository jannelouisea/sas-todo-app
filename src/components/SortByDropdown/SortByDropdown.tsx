import Select from '@mui/joy/Select';
import Option from '@mui/joy/Option';
import PropTypes from 'prop-types';

import { ISortByItem } from 'src/interfaces';
import { Sort } from 'src/static/enums';
import { Label } from 'src/components';
import { useAppDispatch } from 'src/app/hooks';
import { todoItemsSorted } from 'src/features/todoItems/todoItemsSlice';
import { sortByChanged } from 'src/features/todoItemsFilters/todoItemsFiltersSlice';

type Props = {
    className: string
}

function SortByDropdown({ className }: Props) {
    const sortByOptions: ISortByItem[] = [
        {
            value: Sort.Incmpl1st,
            label: 'Incomplete first'
        },
        {
            value: Sort.Cmpl1st,
            label: 'Completed first'
        },
        {
            value: Sort.DateAsc,
            label: 'Created date (Asc)'
        },
        {
            value: Sort.DateDesc,
            label: 'Created date (Desc)'
        },
        {
            value: Sort.TextAsc,
            label: 'Text (A to Z)'
        },
        {
            value: Sort.TextDesc,
            label: 'Text (Z to A)'
        },
    ];

    const dispatch = useAppDispatch();

    const onSelectChange = (evt: React.SyntheticEvent | null, newValue: string | null) => {
        dispatch(sortByChanged(newValue as Sort));
        dispatch(todoItemsSorted(newValue as Sort));
    }

    return (
        <div className='sort-by-dropdown'>
            <Label text='Sort by' />
            <Select
                defaultValue={Sort.Incmpl1st}
                onChange={onSelectChange}
                slotProps={{
                    root: {
                        sx: {
                            minHeight: '2rem',
                            maxHeight: '2rem'
                        }
                    },
                }}
            >
                {
                    sortByOptions.map((option: ISortByItem, index: number) =>
                        <Option key={`sort-by-${index}`} value={option.value}>
                            {option.label}
                        </Option>
                    )
                }
            </Select>
        </div>
    );
}

SortByDropdown.propTypes = {
    className: PropTypes.string,
}

SortByDropdown.defaultProps = {
    className: ''
}

export default SortByDropdown;