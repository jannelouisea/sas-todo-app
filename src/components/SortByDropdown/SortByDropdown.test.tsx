import renderer from 'react-test-renderer';
import SortByDropdown from './SortByDropdown';
import { Provider } from 'react-redux';
import { render, screen, fireEvent } from '@testing-library/react';

import * as todoItemsActions from 'src/features/todoItems/todoItemsSlice';
import * as todoItemsFiltersActions from 'src/features/todoItemsFilters/todoItemsFiltersSlice';
import {
    IStoreType,
    createStore
} from 'src/app/store';
import { Sort } from 'src/static/enums';

describe('SortByDropdown', () => {
    let store: IStoreType;

    beforeEach(() => {
        store = createStore();
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <SortByDropdown className='foo' />
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('saves sorting order and sorts list on changes', () => {
        const sortByChangedSpy = jest.spyOn(todoItemsFiltersActions, 'sortByChanged');
        const todoItemsSortedSpy = jest.spyOn(todoItemsActions, 'todoItemsSorted');

        render(
            <Provider store={store}>
                <SortByDropdown className='foo' />
            </Provider>
        );

        // Verify select dropdown exists
        let select: HTMLInputElement = screen.getByLabelText('Sort by select');
        expect(select).toBeInTheDocument();
        fireEvent.click(select);

        // Simulate selecting an option
        const sortByOption = screen.getByRole("option", {
            name: 'Text (A to Z)'
        });
        expect(sortByOption).toBeInTheDocument();

        fireEvent.click(sortByOption);

        // Verify dispatch functions are called
        expect(sortByChangedSpy).toHaveBeenCalledTimes(1);
        expect(sortByChangedSpy).toHaveBeenCalledWith(Sort.TextAsc);
        expect(todoItemsSortedSpy).toHaveBeenCalledTimes(1);
        expect(todoItemsSortedSpy).toHaveBeenCalledWith(Sort.TextAsc);
    });
});