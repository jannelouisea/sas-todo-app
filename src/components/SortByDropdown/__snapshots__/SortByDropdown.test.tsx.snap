// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortByDropdown renders correctly 1`] = `
<div
  className="sort-by-dropdown foo"
>
  <div
    className=""
  >
    <p
      className="MuiTypography-root MuiTypography-body3 MuiTypography-colorNeutral MuiTypography-variantPlain css-1v5xtsw-JoyTypography-root"
    >
      Sort by
    </p>
  </div>
  <div
    className="MuiSelect-root MuiSelect-variantOutlined MuiSelect-colorNeutral MuiSelect-sizeMd css-dlq8yc-JoySelect-root"
  >
    <button
      aria-controls=":r0:"
      aria-expanded={false}
      aria-label="Sort by select"
      className="MuiSelect-button css-1hr8yhh-JoySelect-button"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      role="combobox"
      tabIndex={0}
    />
    <span
      className="MuiSelect-indicator css-o99lva-JoySelect-indicator"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-colorInherit MuiSvgIcon-fontSizeXl css-hl2kr4-JoySvgIcon-root"
        data-testid="UnfoldIcon"
        focusable={false}
        viewBox="0 0 24 24"
      >
        <path
          d="m12 5.83 2.46 2.46c.39.39 1.02.39 1.41 0 .39-.39.39-1.02 0-1.41L12.7 3.7a.9959.9959 0 0 0-1.41 0L8.12 6.88c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0L12 5.83zm0 12.34-2.46-2.46a.9959.9959 0 0 0-1.41 0c-.39.39-.39 1.02 0 1.41l3.17 3.18c.39.39 1.02.39 1.41 0l3.17-3.17c.39-.39.39-1.02 0-1.41a.9959.9959 0 0 0-1.41 0L12 18.17z"
        />
      </svg>
    </span>
  </div>
</div>
`;
